File: gog/packages/shared/src/log.js
/**
 * Structured logs never include mail/event bodies, queries, tokens, or keyring secrets.
 */
export function logEvent(scope, event, extra = {}) {
  const forbidden = new Set([
    'body',
    'text',
    'note',
    'query',
    'summary',
    'description',
    'raw_json',
    'rawjson',
    'stdout',
    'stderr',
    'argv',
    'args',
    'cookie',
    'cookies',
    'token',
    'password',
    'authorization',
    'caption',
  ]);
  const safe = {};
  for (const [key, value] of Object.entries(extra)) {
    if (forbidden.has(key.toLowerCase())) continue;
    safe[key] = value;
  }
  console.log(JSON.stringify({ ts: Date.now(), scope, event, ...safe }));
}

File: messages/workers/gmessages/internal/log/log.go
package log

import (
	"encoding/json"
	"os"
	"strings"
	"time"
)

var forbidden = map[string]struct{}{
	"body":          {},
	"text":          {},
	"raw_json":      {},
	"rawjson":       {},
	"cookie":        {},
	"cookies":       {},
	"token":         {},
	"password":      {},
	"authorization": {},
	"caption":       {},
}

func Event(scope, event string, extra map[string]any) {
	out := map[string]any{
		"ts":    time.Now().UnixMilli(),
		"scope": scope,
		"event": event,
	}
	for k, v := range extra {
		if _, skip := forbidden[strings.ToLower(k)]; skip {
			continue
		}
		out[k] = v
	}
	enc := json.NewEncoder(os.Stdout)
	_ = enc.Encode(out)
}
